apps/hash-ai-worker-ts/src/activities/flow-activities/research-entities-action/infer-entities-from-web-page-worker-agent.ts

```diff
@@ -108,21 +94,57 @@ const toolDefinitions: Record<ToolId, ToolDefinition<ToolId>> = {
           type: "string",
           description: "The URL of the web page.",
         },
-        text: {
+        hmtlContent: {
           type: "string",
           description: dedent(`
-            The relevant text from the web page from which to infer entities.
+            The HTML content with the relevant sections, paragraphs, tables
+              or other content from the webpage that describe entities of the requested type(s).
 
-            Provide the entire text necessary to accurately infer properties of the
-            relevant entities in a single tool call.
-            
-            Do not truncate or provide partial text which may lead to missed entities
+            You must not modify the content in any way.
+
+            You must include table headers when passing data from a table.
+
+            Do not under any circumstance truncate or provide partial text which may lead to missed entities
               or properties.
-        
-            If additional context (e.g. the current date) is needed to fully infer the
-            entities, ensure it is incorporated in the provided text. The current is ${new Date().toISOString()}.
+
+            You must provide as much text as necessary to infer all
+              the required entities and their properties from the web page in a single tool call.
+
+            ${
+              ""
+              /**
+               * Note: This was previously included to sometimes improve the unit formatting
+               * for the inference agent, however this came with the cost that the worker
+               * agent would frequently truncate text. This has been omitted so that the
+               * unit inference becomes the responsibility of the inference agent.
+               */
+              // For example if the text contains data from a table, you must provide the table
+              // column names. If the are column names specifying units, you must explicitly
+              // specify the unit for each value in the table. For example if the column
+              // specifies a unit in millions (m), append this to each value (e.g. "10 million (m)").
+
+              /** Note: the agent doesn't do this even if you ask it to */
+              // If there are units in the data, you must give a detailed definition for
+              // each unit and what it means.
+            }
             `),
         },
+        expectedNumberOfEntities: {
+          type: "number",
+          description: dedent(`
+            The expected number of entities which should be inferred from the text.
+            You should expect at least 1 entity to be inferred.
+          `),
+        },
```

Comment:
This presumably relies on the LLM using the tool to know exactly how many entities of the expected types are present in the web page. We'll have to see how that works when the research goal doesn't have a clear expectation of how many entities there are to discover, and it's just 'as many as you can'.

Reply:
Hmm that's a fair point, it's worked fine for now (the worker agent is aware of how many entities there are in the content it's sending to the inference agent) - but with different use-cases it might struggle. Maybe we support a `"unknown"` or `"as many as possible"` value for the argument.

Making it optional would mean that the agent will probably not use it, I haven't once had the agent provide an optional argument.

I've added a `@todo` in https://github.com/hashintel/hash/pull/4305/commits/0e9fd993f133eddb6579e56b63b32c79f1969e9f